Move Sorbet gems into an optional group

Bo98 opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Whenever a new macOS version is released, all operations which invoke `Homebrew.install_bundler_gems!` are non-functional until Sorbet releases a gem compatible with the new macOS version.

This PR limits this impact by only installing Sorbet gems when `brew typecheck` or `brew tests` is run. This means that other commands like `brew audit`, `brew style`, `brew test` etc can still be run - enabling formula development to happen.

This also covers unsupported archs. While not officially supported by us, this PR will allow most developer commands to work again on those systems.

The list of groups installed is remembered. So if you run `brew typecheck` once, then it'll keep Sorbet installed and updated when you run `brew audit`, for example.

MikeMcQuaid reviewed (dismissed):
Makes sense to me! My only worry is that we end up uninstalling these in some circumstances due to `BUNDLE_CLEAN`.

Bo98 commented:
In which scenarios would that happen? If it's on `bundle install` then I store the gem groups instored via `Homebrew::Settings.write(:gemgroups, ...)` and retrieve it again when `Homebrew.install_bundler_gems!` is called.

The groups passed into `Homebrew.install_bundler_gems!` append to this list (though there's only one possible value at the moment). It's possible clear this with a `brew install-bundler-gems --groups=` if desired.

MikeMcQuaid commented:
> In which scenarios would that happen? If it's on `bundle install` then I store the gem groups installed via `Homebrew::Settings.write(:gemgroups, ...)` and retrieve it again when `Homebrew.install_bundler_gems!` is called.

Sounds like: no scenarios 🤞🏻.

✅ 🚢

Bo98 commented:
Using `Homebrew::Settings` here is actually a bit of a pain here.

`Homebrew.install_bundler_gems!` is invoked very early in the loading process when `HOMEBREW_BOOTSNAP` or `HOMEBREW_SORBET_RUNTIME` is set. This means that the vast majority of Homebrew isn't loaded at this point.

Challenges are:
* Sorbet type signatures.
  - In the case of `HOMEBREW_SORBET_RUNTIME`, Sorbet might not be installed yet.
  - In the `HOMEBREW_BOOTSNAP` case, even the stubs haven't been loaded yet.
  - Solution would be to move the signatures to RBI files, which makes sense but has a cascading effect - the Settings module relies on `system_command` (though this could be changed to not use that) and also relies on `config.rb` which loads `extend/git_repository`, all of which also use type signatures.
* The use of `HOMEBREW_REPOSITORY`. This env isn't always set in all cases. Running `brew style` with `HOMEBREW_SORBET_RUNTIME` or `HOMEBREW_BOOTSNAP` will invoke `Homebrew.install_bundler_gems!`, but this env isn't set so we need to figure out the repository through other methods.

Basically `Homebrew::Settings` would need to be stripped down to basically pure Ruby with ideally minimal to no `require` lines or other external help.

MikeMcQuaid commented:
> Using `Homebrew::Settings` here is actually a bit of a pain here.

I'd pass on using it given the above. I'm wondering if a much simpler solution would be better e.g. making the `Gemfile` conditional on the OS version/arch?

Bo98 commented:
Ok so I played around with this just to experiment with what approach might be best.

I learned that there were parts of `utils/gems` that already required env variables, so I did the following:

* Moved `config` earlier in the load cycle. This make sure the brew environment is set up and creates the necessary constants.
  - To do this I have a `startup` file which is `load_path` + extras like Bootstrap and config loading. This allows Rubocop to still work outside the Homebrew environment by just using `load_path` (`Homebrew.install_bundler_gems!` is not invoked nor supported in this scenario).
  - This unveiled a couple bugs - like the use of `HOMEBREW_LIBRARY` env in `Utils.setup_gem_environment!` - which will not exist when running under Rubocop (but `HOMEBREW_LIBRARY_PATH` will since that's not reliant on an env).
* Restricted `settings` to use `Kernel.system` (with Ruby 2.6's own exception throwing feature - this didn't exist when we first created `safe_system`) and `Utils.popen_read`. I don't think I've regressed anything here but I realise this takes a different direction from recent changes so have a look and see if it matches what you expect.
* Moved a limited amount of type signatures to RBI files - this is restricted to things that need to load before Sorbet (we do so already for `utils/gems` for similar reasons). `extend/git_repository` is the main change here, which is a little annoying since it is only used for one constant and those methods don't actually matter at the gem loading stage.
* Fixed errors (already present before this PR but only revealed in tests by the above changes) when running `rubocop` outside the brew environment but with either `HOMEBREW_BOOTSNAP` or `HOMEBREW_SORBET_RUNTIME` set.

It seems like a bit of moving around just to get something not too important to work, but since doing this revealed a couple bugs then I'm leaning towards this being the better approach to make.

However, if you disagree I'm happy to just do this:

> making the `Gemfile` conditional on the OS version/arch?

MikeMcQuaid commented on Library/Homebrew/config.rb:
```diff
@@ -12,9 +12,6 @@ class MissingEnvironmentVariables < RuntimeError; end
 module EnvVar
-  extend T::Sig
-
-  sig { params(env: String).returns(String) }
```
What happens if you add this back? Does it error-out in a sufficiently easy to understand version? I wonder if it's worth having a comment here noting the `.rbi` file?

Bo98 replied:
It errors because `require "sorbet-runtime"` or `require "sorbet-runtime-stub"` hasn't been run yet so `T` doesn't exist.

MikeMcQuaid replied:
Yeh, I figured that's _why_ it errored but I was wondering _how_ it errored i.e. if the messaging is obvious enough that people will understand what they've done wrong here.

Bo98 replied:
```
Traceback (most recent call last):
	5: from /usr/local/Homebrew/Library/Homebrew/brew.rb:31:in `<main>'
	4: from /usr/local/Homebrew/Library/Homebrew/brew.rb:31:in `require_relative'
	3: from /usr/local/Homebrew/Library/Homebrew/global.rb:75:in `<top (required)>'
	2: from /usr/local/Homebrew/Library/Homebrew/global.rb:75:in `require'
	1: from /usr/local/Homebrew/Library/Homebrew/config.rb:14:in `<top (required)>'
/usr/local/Homebrew/Library/Homebrew/config.rb:15:in `<module:EnvVar>': uninitialized constant EnvVar::T (NameError)
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/install-bundler-gems.rb:
```diff
@@ -15,14 +15,19 @@ def install_bundler_gems_args
         Install Homebrew's Bundler gems.
       EOS
+      comma_array "--groups=",
+                  description: "Installs the specified comma-separated list of gem groups (default: last remembered)."
```
```suggestion
                  description: "Installs the specified comma-separated list of gem groups (default: last used)."
```

(assuming that's correct?)

MikeMcQuaid commented on Library/Homebrew/extend/git_repository.rb:
```diff
@@ -8,102 +8,84 @@
 # @see Utils::Git
 # @api private
 module GitRepositoryExtension
-  extend T::Sig
```
This feels non-ideal (more than above). Again, warrants a comment perhaps? Also wondering if making `Settings` just not use `GitRepositoryExtension` would be better?

Bo98 replied:
Yeah this is the biggest change tbh and is only because `HOMEBREW_REPOSITORY` uses it. The rest are indeed very trivial.

Bo98 replied:
I could make `HOMEBREW_REPOSITORY` only use it later in the loading cycle - even though messing with constants does seem a little smelly.

MikeMcQuaid replied:
Yeh, I thought the same and agreed it does seem a little smelly.

Bo98 replied:
Alternative is to just ditch `GitRepositoryExtension` for `HOMEBREW_REPOSITORY` and just localise it to where it is needed - which is just a diagnostic and SystemConfig.

MikeMcQuaid replied:
Yeh, that makes more sense to me 👍🏻

MikeMcQuaid commented on Library/Homebrew/utils/gems.rb:
```diff
@@ -128,23 +128,33 @@ def install_bundler!
 
+    require "settings"
+
+    groups |= (Homebrew::Settings.read(:gemgroups)&.split(";") || [])
```
This is a little opaque; a comment might be nice.

MikeMcQuaid commented on Library/Homebrew/utils/gems.rb:
```diff
@@ -128,23 +128,33 @@ def install_bundler!
-    @bundle_installed ||= begin
+    ENV["BUNDLE_WITH"] = groups.join(" ")
+
+    if @bundle_installed_groups != groups
```
Could this be an early `return`?

Bo98 replied:
There is a `setup_gem_environment!` below this.

MikeMcQuaid commented on Library/Homebrew/utils/rubocop.rb:
```diff
@@ -2,9 +2,7 @@
-require_relative "gems"
-Homebrew.setup_gem_environment!
-
+require_relative "../startup"
```
Why does `rubocop` need this, out of interest?

Bo98 replied:
It requires the warning gem so we should be setting up the gem environment and the load path properly to use it.

This probably could have been `../load_path` - I'll double check.

Note that this file (unlike `rubocops.rb`) is only invoked with Rubocop runs via `brew`.

MikeMcQuaid replied:
> This probably could have been `../load_path` - I'll double check.

That'd be cool if so 👍🏻

MikeMcQuaid commented on Library/Homebrew/utils/sorbet.rb:
```diff
@@ -1,8 +1,9 @@
-if ENV["HOMEBREW_SORBET_RUNTIME"]
-  Homebrew.install_bundler_gems!
+if ENV["HOMEBREW_SORBET_RUNTIME"] && ENV["HOMEBREW_BREW_FILE"]
+  # This is only supported under the brew environment.
```
When else is this run?

Bo98 replied:
If you run rubocop outside of `brew`, it might have invoked gem installation if you happen to have `HOMEBREW_SORBET_RUNTIME` in your environment (which is a reasonable scenario to happen).

This limits it to only run under the brew environment - `HOMEBREW_BREW_FILE` should never be set by the user (we depend on that in other places too).

Bo98 replied:
`utils/sorbet` is a require in `rubocops.rb`.

MikeMcQuaid replied:
Gotcha. I kinda wonder if we should have separate directories for:
- stuff that we expect to be run when `HOMEBREW_BREW_FILE` is unset (e.g. used by `rubocops.rb`): maybe `standalone/`?
- stuff like what you've made above that's needed at `brew.rb` boot before e.g. `global.rb` is required: maybe `startup/`?
 
I think that'd make some of these implicit dependencies a bit more obvious. Thoughts?

Bo98 replied:
Currently the scenario is:

- Standalone: `load_path` and `utils/sorbet`
- Startup: `load_path`, `startup.rb` (which loads `config.rb` and Bootsnap) and `utils/sorbet`.

Splitting `utils/sorbet` is probably a good idea - it's just a bit annoying that `require "sorbet-runtime-stub"` needs hacks to avoid `sorbet-runtime` (but maybe we shouldn't be caring in the standalone case? Just let it happen?).

MikeMcQuaid replied:
> * Standalone: `load_path` and `utils/sorbet`
> * Startup: `load_path`, `startup.rb` (which loads `config.rb` and Bootsnap) and `utils/sorbet`.

I guess I'd propose:
- `standalone.rb`, `standalone/load_path`, `standalone/sorbet`
- `startup.rb`, `startup/config`, `startup/bootsnap`

?

> Splitting `utils/sorbet` is probably a good idea - it's just a bit annoying that `require "sorbet-runtime-stub"` needs hacks to avoid `sorbet-runtime` (but maybe we shouldn't be caring in the standalone case? Just let it happen?).

Either sounds good to me.

MikeMcQuaid reviewed (commented):
> However, if you disagree I'm happy to just do this:
>> making the Gemfile conditional on the OS version/arch?

I'm not sure I feel super strongly in one direction or another but it might be worth playing around with that (time boxed, please don't let me waste too much more of your time) in another PR so we can compare the two. The amount of changes required here just to not install Sorbet are starting to feel a little heavy (but perhaps for the best, hard to tell without comparing to another approach)? 

Feel free to push back on this or any other requests, too.

MikeMcQuaid reviewed (approved):
Yeh, this looks great, happy with this now. Nice work @Bo98!

Bo98 commented:
Awesome.

This provides a framework for expansion should it be needed for anything else, for example a `man` group for `brew man` dependencies.

I don't have any particular personal need for grouping beyond Sorbet, but install speed (less to install), install reliability (less gems to compile and potentially break), and limiting exposure of what gems are available in CI for `brew test <formula>` are all possible reasons to further groups.

MikeMcQuaid commented:
> Whenever a new macOS version is released, all operations which invoke `Homebrew.install_bundler_gems!` are non-functional until Sorbet releases a gem compatible with the new macOS version.

@bo98 this is applicable now so it'd be great if you could get CI green here when you get a chance 💚

Bo98 commented:
Everything works fine for me locally.

> undefined method `error_message_for_obj' for nil:NilClass

Not sure why not in the CI.

MikeMcQuaid commented:
> Everything works fine for me locally.
> 
> > undefined method `error_message_for_obj' for nil:NilClass
> 
> Not sure why not in the CI.

@Bo98 perhaps try to unset `HOMEBREW_DEVELOPER` locally and start from a clean state with `git clean -xdf` in your `brew --repo`?

Bo98 commented:
Still can't reproduce. I suspect it is some parallelism or spec ordering issue.

Manually trawling through the code, this line could be problematic:

https://github.com/Homebrew/brew/blob/33e71c99dc40c0a134a236fffe2594ae6572d83c/Library/Homebrew/style.rb#L88

since it will trigger the load path and Sorbet to be reinitialised (`rubocops.rb` is designed to be run standalone).

Bo98 commented:
Apart from macOS coverage being stuck as "queued for processing" and thus showing an overall -5.61% (which is probably a serverside issue and not to do with my changes), this seems to work now.

Bo98 commented:
Lastly, this doesn't do anything anymore:

https://github.com/Homebrew/brew/blob/f7d54023c34becec896ce7cb16230114b1644109/Library/Homebrew/test/support/lib/startup/config.rb#L52-L59

because this is now run before default prefix in global.rb is established and thus won't be overriding anything.

Do we need it? Only difference I noticed is a different execution path in dev-cmd/bottle, which doesn't break tests.
If it's desired to keep, I suppose we can have a separate `default_prefix.rb` or something.

Bo98 commented:
> If it's desired to keep, I suppose we can have a separate `default_prefix.rb` or something.

Did this to be safe.

MikeMcQuaid commented:
Looks good, thanks @Bo98!
